scylla-cdc/src/cdc_reader.rs

```diff
@@ -0,0 +1,213 @@
+use std::sync::Arc;
+
+use anyhow;
+use chrono::Duration;
+use core::time;
+use futures::future::RemoteHandle;
+use futures::stream::{FusedStream, FuturesUnordered, StreamExt};
+use futures::FutureExt;
+use scylla::Session;
+use tokio::task::JoinError;
+
+use crate::cdc_types::GenerationTimestamp;
+use crate::consumer::ConsumerFactory;
+use crate::reader::StreamReader;
+use crate::stream_generations::GenerationFetcher;
+
+pub struct CDCReader {
+    // Tells the worker to stop
+    // Usage of the "watch" channel will make it possible to change the the timestamp later,
+    // for example if somebody loses patience and wants to stop now not later
+    end_timestamp: tokio::sync::watch::Sender<Duration>,
+}
+
+impl CDCReader {
+    // Creates a reader and runs it immediately
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        session: &Arc<Session>,
+        keyspace: String,
+        table_name: String,
+        start_timestamp: Duration,
+        end_timestamp: Duration,
+        window_size: Duration,
+        safety_interval: Duration,
+        sleep_interval: time::Duration,
+        consumer_factory: Box<dyn ConsumerFactory + Send>,
+    ) -> (Self, RemoteHandle<Result<anyhow::Result<()>, JoinError>>) {
+        let (end_timestamp_sender, end_timestamp_receiver) =
+            tokio::sync::watch::channel(end_timestamp);
+        let mut worker = CDCReaderWorker::new(
+            session,
+            keyspace,
+            table_name,
+            start_timestamp,
+            end_timestamp,
+            window_size,
+            safety_interval,
+            sleep_interval,
+            end_timestamp_receiver,
+            consumer_factory,
+        );
+        let (_, handle) = tokio::task::spawn(async move { worker.run().await }).remote_handle();
+        let printer = CDCReader {
```

Comment:
This piece of code (spawning) required the `ConsumerFactory` object to implement also a `Send` trait. From what I've seen, it's probably automatically generated if all members of a structure implement Send, so it's probably not a bothering issue, is it?

Reply:
Yes, requiring `Send` is fine. However, I think that `ConsumerFactory` should require `Sync`, which is a stronger trait than `Send`. This will allow the factory to be shared by multiple tasks and you will be able to keep `ConsumerFactory::new_consumer` async and make it possible to call it concurrently.